File: test-processor.py
#!/usr/bin/env python3
"""
Contour-based processor: trace original stroke paths, smooth with splines
"""

import cv2
import numpy as np
import json
from pathlib import Path
from scipy.interpolate import splprep, splev


def process_image(image_path):
    """Process handwriting using contour approximation"""
    
    img = cv2.imread(str(image_path))
    if img is None:
        print(f"❌ Could not read {image_path}")
        return []
    
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)
    _, binary = cv2.threshold(blurred, 180, 255, cv2.THRESH_BINARY_INV)
    
    kernel = np.ones((3, 3), np.uint8)
    binary = cv2.morphologyEx(binary, cv2.MORPH_CLOSE, kernel)
    binary = cv2.morphologyEx(binary, cv2.MORPH_OPEN, kernel)
    
    contours, _ = cv2.findContours(binary, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if not contours:
        print(f"⚠️  No content detected in {image_path.name}")
        return []
    
    filtered_contours = [cnt for cnt in contours if cv2.contourArea(cnt) > 100]
    if not filtered_contours:
        print(f"⚠️  No significant content detected in {image_path.name}")
        return []
    
    x, y, w, h = cv2.boundingRect(np.vstack(filtered_contours))
    
    padding = 20
    x = max(0, x - padding)
    y = max(0, y - padding)
    w = min(gray.shape[1] - x, w + 2 * padding)
    h = min(gray.shape[0] - y, h + 2 * padding)
    
    cropped_gray = gray[y:y+h, x:x+w]
    cropped_binary = binary[y:y+h, x:x+w]
    
    max_dim = max(w, h)
    square = np.ones((max_dim, max_dim), dtype=np.uint8) * 255
    y_offset = (max_dim - h) // 2
    x_offset = (max_dim - w) // 2
    square[y_offset:y_offset+h, x_offset:x_offset+w] = cropped_gray
    
    _, square_binary = cv2.threshold(square, 180, 255, cv2.THRESH_BINARY_INV)
    
    contours, _ = cv2.findContours(square_binary, cv2.RETR_LIST, cv2.CHAIN_APPROX_TC89_L1)
    
    print(f"  - Found {len(contours)} contour(s)")
    
    all_points = []
    
    for i, contour in enumerate(contours):
        if len(contour) < 10:
            continue
        
        epsilon = 0.01 * cv2.arcLength(contour, False)
        approx = cv2.approxPolyDP(contour, epsilon, False)
        
        points = [(pt[0][0], pt[0][1]) for pt in approx]
        
        if len(points) >= 4:
            smoothed = smooth_path_with_spline(points, smoothness=5, num_points=50)
            all_points.extend(smoothed)
            print(f"  - Contour {i+1}: {len(contour)} → {len(approx)} → {len(smoothed)} points")
        else:
            all_points.extend(points)
            print(f"  - Contour {i+1}: {len(contour)} → {len(points)} points (too short to smooth)")
    
    normalized = [[x / max_dim, y / max_dim] for x, y in all_points]
    
    print(f"✓ Processed {image_path.name}: {len(normalized)} points")
    return normalized


def smooth_path_with_spline(points, smoothness=0, num_points=50):
    """Smooth path using cubic spline interpolation"""
    if len(points) < 4:
        return points
    
    points = np.array(points)
    x = points[:, 0]
    y = points[:, 1]
    
    try:
        tck, u = splprep([x, y], s=smoothness, k=min(3, len(points)-1))
        u_new = np.linspace(0, 1, num_points)
        x_new, y_new = splev(u_new, tck)
        return list(zip(x_new, y_new))
    except:
        return points


def main():
    print("\n🚀 Processing test images...\n")
    
    test_dir = Path(__file__).parent / "test-data"
    characters = ['A']
    image_files = ['A.png']
    
    glyphs = []
    
    for char, filename in zip(characters, image_files):
        image_path = test_dir / filename
        
        if not image_path.exists():
            print(f"⚠️  File not found: {image_path}")
            continue
        
        print(f'Processing "{char}" from {filename}...')
        points = process_image(image_path)
        
        if points:
            glyphs.append({
                "char": char,
                "strokes": [points]
            })
    
    if not glyphs:
        print("\n❌ No glyphs processed.\n")
        return
    
    output = {
        "fontName": f"test_{Path(__file__).stem}",
        "glyphs": glyphs
    }
    
    output_path = test_dir / "test-output.json"
    with open(output_path, 'w') as f:
        json.dump(output, f, indent=2)
    
    print(f"\n✅ Success! Processed {len(glyphs)} character(s)")
    print(f"📄 Output saved to: {output_path}\n")


if __name__ == "__main__":
    main()

File: test-processor.js
const sharp = require('sharp');
const fs = require('fs').promises;
const path = require('path');

// Character mapping
const CHARACTERS = ['!', 'A', '0'];

async function processImage(imagePath) {
  try {
    const image = sharp(imagePath);
    const metadata = await image.metadata();
    const width = metadata.width || 1000;
    const height = metadata.height || 1000;

    // Process: grayscale -> threshold -> negate
    const processed = await image
      .grayscale()
      .threshold(128)
      .negate()
      .raw()
      .toBuffer();

    // Extract points
    const points = extractPoints(processed, width, height);
    
    if (points.length === 0) {
      console.log(`⚠️  No handwriting detected in ${path.basename(imagePath)}`);
      return [];
    }

    // Normalize coordinates
    const normalized = points.map(p => ({
      x: p.x / width,
      y: p.y / height,
    }));

    console.log(`✓ Processed ${path.basename(imagePath)}: ${points.length} points extracted`);
    return normalized;
  } catch (error) {
    console.error(`✗ Error processing ${imagePath}:`, error.message);
    return [];
  }
}

function extractPoints(data, width, height) {
  const points = [];
  const samplingDistance = 5;
  
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = y * width + x;
      const pixel = data[idx];
      
      // If black pixel (handwriting)
      if (pixel > 128) {
        // Sample every Nth point to reduce data size
        if (points.length === 0) {
          points.push({ x, y });
        } else {
          const last = points[points.length - 1];
          const distance = Math.sqrt(
            Math.pow(x - last.x, 2) + Math.pow(y - last.y, 2)
          );
          
          if (distance >= samplingDistance) {
            points.push({ x, y });
          }
        }
      }
    }
  }
  
  return sortByProximity(points);
}

function sortByProximity(points) {
  if (points.length <= 1) return points;
  
  const sorted = [];
  const remaining = [...points];
  
  sorted.push(remaining.shift());
  
  while (remaining.length > 0) {
    const last = sorted[sorted.length - 1];
    let nearestIndex = 0;
    let nearestDistance = Infinity;
    
    for (let i = 0; i < remaining.length; i++) {
      const distance = Math.sqrt(
        Math.pow(remaining[i].x - last.x, 2) +
        Math.pow(remaining[i].y - last.y, 2)
      );
      
      if (distance < nearestDistance) {
        nearestDistance = distance;
        nearestIndex = i;
      }
    }
    
    sorted.push(remaining.splice(nearestIndex, 1)[0]);
  }
  
  return sorted;
}

async function main() {
  console.log('\n🚀 Processing test images...\n');

  const testDir = path.join(__dirname, 'test-data');
  const imageFiles = [
    'exclamation.png',
    'A.png',
    'zero.png'
  ];

  const glyphs = [];

  for (let i = 0; i < imageFiles.length; i++) {
    const imagePath = path.join(testDir, imageFiles[i]);
    const char = CHARACTERS[i];
    
    console.log(`Processing "${char}" from ${imageFiles[i]}...`);
    
    try {
      await fs.access(imagePath);
    } catch {
      console.log(`⚠️  File not found: ${imagePath}`);
      console.log(`   Please create this file first.\n`);
      continue;
    }

    const points = await processImage(imagePath);
    
    if (points.length > 0) {
      // Convert to stroke format (array of [x,y] pairs)
      const stroke = points.map(p => [p.x, p.y]);
      
      glyphs.push({
        char: char,
        strokes: [stroke]
      });
    }
  }

  if (glyphs.length === 0) {
    console.log('\n❌ No glyphs processed. Make sure image files exist in test-data/\n');
    return;
  }

  // Create JSON output
  const output = {
    fontName: `test_${new Date().toISOString().split('T')[0]}`,
    glyphs: glyphs
  };

  // Save to file
  const outputPath = path.join(testDir, 'test-output.json');
  await fs.writeFile(outputPath, JSON.stringify(output, null, 2));

  console.log(`\n✅ Success! Processed ${glyphs.length} character(s)`);
  console.log(`📄 Output saved to: ${outputPath}\n`);
  console.log('Preview:');
  console.log(JSON.stringify(output, null, 2).substring(0, 500) + '...\n');
}

main().catch(console.error);
